design/user-event-spi.md

```diff
@@ -0,0 +1,149 @@
+# User Event SPI
+
+* **Status**: Draft v0.1
+* **Author**: [vmuzikar](https://github.com/vmuzikar)
+
+## Motivation
+
+Current Event SPI design has several limitations. E.g. it's not defined when an event is fired exactly, there are no pre-
+and post-events. There's also no way of pre-events intercepting for approve/reject functionality.
+
+## Use cases
+
+* A user is removed from Keycloak. All related records in an external system are removed before the user is permanently
+  deleted (e.g. eshop orders).
+* A user self-registers in Keycloak. Their registration must be approved by an external system (either human or automated)
+  before the new account is actually fully created.
+* A user self-edits their email address. The change must be approved first by an external system before it's actually
+  applied.
+
+## Design proposal
+
+To overcome the current Event SPI limitations, a new "domain-specific SPI" will be implemented – User Event SPI.
+
+Key changes over current generic Event SPI:
+* Each event is fired twice – as a pre- and post-event.
+* Pre-events are "interactive" – listener can approve, reject or delegate (for later decision by an external system)
+  each pre-event. 
+* The same User Event SPI will be used everywhere – admin console and REST endpoints, user registration form,
+  account console and REST endpoints, etc.
+* The event types will be much more detailed, not just usual CRUD. There'll be separated events for user self-registration,
+  creation through Admin REST API, Account Console, etc.
+
+### Terminology
+
+User Action (in this context) is a process which fires the User Event. This can be e.g. a user creation through Admin REST
+API, user self-registration, user field modification, assigning a user role, etc.
+
+Pre-event is an event that is fired after all other verifications passed (authorization, request validation, etc.), just
+before persisting any changes.
+
+Post-event is an event that is fired after a User Action has finished completely and all changes are persisted.
+
+### Event types
+
+* User is created manually through admin endpoints
+* Basic user info (name, email, etc.) is modified through admin endpoints
+* Basic user info (name, email, etc.) is modified through account endpoints
+* User self-registers through login page
```

Comment:
I would just add the event type of `UserCreated`. And then add a `source` field inside.

From a Microservices perspective, in like 90% of the cases, it's irrelevant where user came from, so the `source` should be a complementary field and not a specific event